Add hid manager definitions

forticulous opened the pull request:
This adds some hid manager functions and types. Take a look and tell me if you think it would make a good addition

Commit:
Add hid manager definitions

dcuddeback commented:
@forticulous Hi. Yes, this looks like it belongs here. Thanks for submitting it. I'll try to find some time later this week to double-check the definitions against the IOKit headers and merge.

forticulous commented:
Did you get a chance to double-check the function definitions?

dcuddeback commented:
@forticulous Sorry, I forgot. I'll look at it now.

dcuddeback commented on src/hid_manager.rs:
```diff
@@ -0,0 +1,42 @@
+pub struct __IOHIDManager {
+    __private: c_void,
+}
+pub type IOHIDManagerRef = *const __IOHIDManager;
```
This should probably be `*mut`. The C typedef doesn't specify `const`:

```c
typedef struct __IOHIDManager * IOHIDManagerRef;
```

And it's passed to functions that most likely mutate, e.g., `IOHIDManagerSetProperty()`.

dcuddeback commented on src/hid_manager.rs:
```diff
@@ -0,0 +1,42 @@
+pub struct __IOHIDDevice {
+    __private: c_void,
+}
+pub type IOHIDDeviceRef = *const __IOHIDDevice;
```
This should probably be `*mut` as well. This is defined in a separate header file (`IOKit/hid/IOHIDBase.h`) than `IOHIDManagerRef` (`IOKit/hid/IOHIDManager.h`). There's a note under the ["contributing" section of the README](https://github.com/dcuddeback/iokit-sys#contributing) that explains the code organization:

> If you plan to submit a pull request, please note the structure of the code. There is one file for each header file in the IOKit framework. For example, `src/io_return.rs` contains the definitions from `IOKit/IOReturn.h`. The definitions in each file are more or less in the same order that they appear in the matching header file. Each file is then re-exported in the crate root, e.g., `pub use io_return::*`.

Would you be able to split the definitions to follow this convention? If you look at the existing files, most of them have a comment at the top saying which header file they correspond to. That makes it easier to find the C definitions for comparing to the Rust defintions.

dcuddeback commented on src/hid_manager.rs:
```diff
@@ -0,0 +1,42 @@
+pub type IOHIDReportCallback = extern fn(context: *mut c_void, result: IOReturn, sender: *mut c_void, report_type: IOHIDReportType, 
+                                         report_id: u32, report: *mut u8, report_len: CFIndex);
+
+pub type IOHIDReportType = u32;
```
The C definition is a plain enum. I think the equivalent type in Rust would be `c_int`.

dcuddeback commented on src/hid_manager.rs:
```diff
@@ -0,0 +1,42 @@
+pub const kIOHIDReportTypeFeature: IOHIDReportType = 2;
+pub const kIOHIDReportTypeCount: IOHIDReportType   = 3;
+
+pub const kIOHIDManagerOptionNone: IOOptionBits                    = 0;
```
I see this defined as part of the `IOHIDManagerOptions` enum. Is this the same place you're pulling the definitions from? (`IOKit/hid/IOHIDManager.h`):

```c
typedef enum {
    kIOHIDManagerOptionNone = 0x0,
    kIOHIDManagerOptionUsePersistentProperties = 0x1,
    kIOHIDManagerOptionDoNotLoadProperties = 0x2,
    kIOHIDManagerOptionDoNotSaveProperties = 0x4,
} IOHIDManagerOptions;
```

dcuddeback replied:
Would be nice to include a typedef for `IOHIDManagerOptions` like with the other enums. I think it makes sense to include `kIOHIDManagerOptionDoNotLoadProperties` and `kIOHIDManagerOptionDoNotSaveProperties` as well as long as we're providing definitions for `IOHIDManagerOptions`.

Commit:
Separate code by header file and other code review changes

forticulous commented:
I made some changes based on your comments. I split the code up by header file and changed the names a little `io_hid_keys.rs` for `IOKit/hid/IOHIDKeys.h` etc. Also made the enum changes. Hopefully its better

Commit:
Removed duplicate code

dcuddeback commented on src/io_hid_manager.rs:
```diff
@@ -0,0 +1,24 @@
+pub type IOHIDManagerRef = *mut __IOHIDManager;
+
+extern "C" {
+    pub fn IOHIDManagerCreate(allocator: CFAllocatorRef, options: IOHIDManagerOptions) -> IOHIDManagerRef;
```
`options` is defined as `IOOptionBits` in the header file I'm looking at:

```c
CF_EXPORT 
IOHIDManagerRef IOHIDManagerCreate(     
                                CFAllocatorRef                  allocator,
                                IOOptionBits                    options)
AVAILABLE_MAC_OS_X_VERSION_10_5_AND_LATER;
```

`IOOptionBits` is a typedef for `u32`, while `IOHIDManagerOptions` is a typedef for `c_int`, which isn't guaranteed to be the same size. Since Rust is stricter than C about integer conversions, I'm guessing that you changed the type here to prevent needing to type `as IOOptionBits` over and over. If this is the only place that `IOHIDManagerOptions` is intended to be used, I think it would make sense to change the target of the `IOHIDManagerOptions` typedef to `IOOptionBits`, since the C compiler would make the same conversion automatically.

dcuddeback commented on src/io_hid_manager.rs:
```diff
@@ -0,0 +1,24 @@
+    pub fn IOHIDManagerRegisterDeviceRemovalCallback(manager: IOHIDManagerRef, callback: IOHIDDeviceCallback, context: *mut c_void);
+    pub fn IOHIDManagerRegisterInputReportCallback(manager: IOHIDManagerRef, callback: IOHIDReportCallback, context: *mut c_void);
+    pub fn IOHIDManagerOpen(manager: IOHIDManagerRef, options: IOHIDManagerOptions) -> IOReturn;
+    pub fn IOHIDManagerClose(manager: IOHIDManagerRef, options: IOHIDManagerOptions) -> IOReturn;
```
Same thing for `IOHIDManagerOpen()` and `IOHIDManagerClose()`. `options` is an `IOOptionBits`, but the `IOHIDManagerOptions` typedef could be changed to avoid explicit casts.

dcuddeback commented on src/io_hid_keys.rs:
```diff
@@ -0,0 +1,15 @@
+pub const kIOHIDReportTypeFeature: IOHIDReportType = 2;
+pub const kIOHIDReportTypeCount: IOHIDReportType   = 3;
+
+pub type IOHIDManagerOptions = c_int;
```
I see this definition in `IOKit/hid/IOHIDManager.h`. I may not have the latest SDK, so just let me know if this has changed since the version I'm looking at.

dcuddeback commented:
@forticulous Thanks for making those changes. I think my earlier comment about enums being like `c_int` may have led you astray on `IOHIDManagerOptions`.

Commit:
Change enum typedef

forticulous commented:
I changed the enum typedefs from `c_int` to `IOOptionBits`

emberian commented:
ping?

dcuddeback commented on src/io_hid_keys.rs:
```diff
@@ -0,0 +1,15 @@
+
+use types::IOOptionBits;
+
+pub type IOHIDReportType = IOOptionBits;
```
I think this one needs to be `c_int`. `IOHIDReportType` is used in the function signature for `IOHIDReportCallback`, so changing this type could affect the ABI, which is dangerous.

dcuddeback commented:
@forticulous My previous suggestion about changing the enum typedef only applied to `IOHIDManagerOptions` enum to mimic C's type promotion. Changing `IOHIDReportType` could break the ABI where it's used in function signatures. Also, note that my previous comments about `options` being `IOOptionBits` in the function signatures still apply.

Commit:
Changing an enum typedef and hid manager function signatures

forticulous commented:
Latest changes:
Changed typedef for `IOHIDReportType` back to `c_int`
Changed the options args in the hid manager functions to `IOOptionBits`

Let me know if i missed anything

dcuddeback commented on src/io_hid_keys.rs:
```diff
@@ -0,0 +1,16 @@
+pub const kIOHIDReportTypeFeature: IOHIDReportType = 2;
+pub const kIOHIDReportTypeCount: IOHIDReportType   = 3;
+
+pub type IOHIDManagerOptions = IOOptionBits;
```
This is defined in `IOHIDManager.h`, not `IOHIDKeys.h`. I think you may have missed this comment from my previous review:

> I see this definition in `IOKit/hid/IOHIDManager.h`. I may not have the latest SDK, so just let me know if this has changed since the version I'm looking at.

I'd like to keep this consistent with the structure that I explained in the ["contributing" section of the README](https://github.com/dcuddeback/iokit-sys#contributing).

dcuddeback commented on src/io_hid_keys.rs:
```diff
@@ -0,0 +1,16 @@
+pub const kIOHIDManagerOptionNone: IOHIDManagerOptions                    = 0;
+pub const kIOHIDManagerOptionUsePersistentProperties: IOHIDManagerOptions = 1;
+pub const kIOHIDManagerOptionDoNotLoadProperties: IOHIDManagerOptions     = 2;
+pub const kIOHIDManagerOptionDoNotSaveProperties: IOHIDManagerOptions     = 4;
```
Since these are intended to be bitflags, I think it would help communicate their intention by defining them as hex constants, similar to how they're defined in `IOHIDManager.h`:

```c
typedef enum {
    kIOHIDManagerOptionNone = 0x0,
    kIOHIDManagerOptionUsePersistentProperties = 0x1,
    kIOHIDManagerOptionDoNotLoadProperties = 0x2,
    kIOHIDManagerOptionDoNotSaveProperties = 0x4,
} IOHIDManagerOptions;
```

I don't have a really strong opinion about this, but it'd be nice to do if you're already moving the definitions to another file.

Commit:
Move IOHIDManagerOptions to correct place, make bitflags hex constants

forticulous commented:
@dcuddeback You're right that IOHIDManagerOptions shouldn't have been in that file. I moved it to io_hid_manager.rs and made the bitflag constants hex literals

dcuddeback commented:
>  I moved it to io_hid_manager.rs

@forticulous Have another look. Looks like you copied them instead of moving them.

Commit:
Remove duplicate code

forticulous commented:
@dcuddeback Oops, totally missed that

dcuddeback commented on src/io_hid_keys.rs:
```diff
@@ -0,0 +1,9 @@
+pub const kIOHIDReportTypeInput: IOHIDReportType   = 0x0;
+pub const kIOHIDReportTypeOutput: IOHIDReportType  = 0x1;
+pub const kIOHIDReportTypeFeature: IOHIDReportType = 0x2;
+pub const kIOHIDReportTypeCount: IOHIDReportType   = 0x3;
```
This enum isn't a bitflag. It's defined in the header like this:

```c
enum IOHIDReportType{
    kIOHIDReportTypeInput = 0,
    kIOHIDReportTypeOutput,
    kIOHIDReportTypeFeature,
    kIOHIDReportTypeCount
};
```

My prior comment was targetted at `IOHIDManagerOptions` only.

dcuddeback committed:
Fix enum values

dcuddeback commented:
@forticulous Thanks for updating. Looks like you also updated an enum to use hex values. I was only suggesting to update the bitflags. I took care of changing the enum back, so that we can move on.

For future reference, please look at the framework's header files and do the simplest translation from what's written in the header file to Rust. I'm comparing the Rust code against the header files when I review. If you do the same before submitting a PR, there's a better chance the PR will go through on the first or second attempt.
